### ISSUE TITLE
false-positive: invalid-format-index

### ISSUE THREAD

**Issue by navidof** [PyCQA/pylint#1537]
### Steps to reproduce
1. pylint test.py 
where test.py contains:
```python
""" Pylint false positive """

class Test:
    """ Triggered when in a class """

    def __init__(self):
        """ init """
        self.the_dict = {}

    def set_dict(self):
        """ doc """
        self.the_dict = {'foo': 1, 'bar': 2}

    def print_dict(self):
        """ print """
        print("{0[foo]} {0[bar]}".format(self.the_dict))
```

### Current behavior
No config file found, using default configuration
************* Module test
W: 16,14: Using invalid lookup key 'foo' in format specifier "0['foo']" (invalid-format-index)
W: 16,14: Using invalid lookup key 'bar' in format specifier "0['bar']" (invalid-format-index)

### Expected behavior
No warnings

### pylint --version output
% pylint --version                                
No config file found, using default configuration 
pylint 1.6.4,                                     
astroid 1.4.9                                     
Python 3.5.2 (default, Jan 19 2017, 12:11:08)     
[GCC 4.8.3 20140911 (Red Hat 4.8.3-7)]

**Comment by PCManticore** [comment]
Thank you for reporting this issue.

**Comment by PCManticore** [comment]
Took a look at this. This happens because pylint tries to see if the formatted element has the attributes/keys that you used in the format string. In this case, it doesn't have them, because you are setting them through the ``set_dict`` method. Unfortunately, pylint has no way to know when ``set_dict`` method is called, thus it assumes these keys are not present, leading to this error.
